dask_expr/reductions.py

```diff
@@ -346,7 +346,10 @@ class Len(Reduction):
     reduction_aggregate = sum
 
     def _simplify_down(self):
-        if isinstance(self.frame, Elemwise):
+        _lengths = Lengths(self.frame).optimize()
+        if isinstance(_lengths, Literal):
+            return Literal(sum(_lengths.value))
+        elif isinstance(self.frame, Elemwise):
             child = max(self.frame.dependencies(), key=lambda expr: expr.npartitions)
             return Len(child)
```

Comment:
It may be that there are cases where we can move Len but not Lengths.  For example:

```python
df = dx.read_csv(...)
df = df.set_index("...")
len(df)
```

I'm not going to be able to pass `Lengths` through the `set_index` call, but I can pass `Len` through (the number of rows is the same after the full shufle).  Because of this I think that they probably need to remain separate operations with separate optimization paths.  I don't think that we can define one in terms of the other.

Also, I suspect that `Len` is likely to be more common than `Lengths`, so I'm disinclined to have it inherit any weaknesses from the other.

Reply:
I think it would be better to keep the `Len(Elemwise)` branch of this on top.  Otherwise we're constructing a `Lengths` object and optimizing it every pass through optimizations (and there are likely to be several of these).  

In principle, I wouldn't mind removing `Lengths` from this entirely, and instead have a `ReadParquet._simplify_up(Len)` case which is similar to but simpler than the `Lengths` case.